### ISSUE TITLE
Unable to decode event `Instruction: X`

### ISSUE THREAD

**Issue by arijoon** [coral-xyz/anchor#2827]
When using event parsers `parseLogs` if `errorOnDecodeFailure` is set to `true` the parser throws on the very first log which is always `Program logged: "Instruction: X"`, `X` being whatever instruction is being executed. Is this an intended feature?

Additionally is it possible for another program to spoof the logs pretending to be a different program we're interested in?

**Comment by acheroncrypto** [comment]
> When using event parsers `parseLogs` if `errorOnDecodeFailure` is set to `true` the parser throws on the very first log which is always `Program logged: "Instruction: X"`, `X` being whatever instruction is being executed. Is this an intended feature?

It's most likely not the intended behavior if it always throws.

> Additionally is it possible for another program to spoof the logs pretending to be a different program we're interested in?

We're essentially parsing the logs so I think it's possible to spoof it, but if you'd like to avoid that, there is `event-cpi` feature which requires "signing" an invocation with a PDA.

**Comment by arijoon** [comment]
> We're essentially parsing the logs so I think it's possible to spoof it, but if you'd like to avoid that, there is `event-cpi` feature which requires "signing" an invocation with a PDA.

Oh thanks for this. I saw that feature but based on original docs assumed its only used to guarantee its available when logs are truncated. If that is far more secure then would certainly opt for it. Assuming it can be parsed with the Anchor event parser in a similar way or it needs to be parsed differently?

**Comment by acheroncrypto** [comment]
It needs to be parsed differently because it's not log based like the normal events. Here is an example:

https://github.com/coral-xyz/anchor/blob/216b56e26f5080ec652b098849e177ec560d602f/tests/events/tests/events.ts#L58-L62

**Comment by arijoon** [comment]
> It needs to be parsed differently because it's not log based like the normal events. Here is an example:

Thanks a lot this is much appreciated, indeed the event in the tx data looks a lot more secure than just parsing logs. One final question. Is checking the innerInstructions and ensuring `account` and `programId` are correct, then parsing the data enough to guarantee no other program could have spoofed this? Basically wondering if any other program and cause this cpi into our program (which has the event) without going through the top level instructions and somehow emit this CPI inner tx?

**Comment by acheroncrypto** [comment]
> Is checking the innerInstructions and ensuring `account` (`__event_authority` of our program) and `programId` are correct, then parsing the data enough to guarantee no other program could have spoofed this? 

Yes, checking those accounts should be enough because the event authority needs to "sign" to make self-CPI work.

> Basically wondering if any other program can cause this cpi into our program (which has the event) without going through the top level instructions and somehow emit this CPI inner tx?

They shouldn't be able to do this because of the same reason above, the event authority needs to "sign" the invocation. We also have a test for this case: https://github.com/coral-xyz/anchor/blob/216b56e26f5080ec652b098849e177ec560d602f/tests/events/tests/events.ts#L69-L101
